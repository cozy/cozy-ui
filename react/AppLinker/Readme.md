Render-props component that provides onClick/href handler to
apply to an anchor that needs to open an app.

If the app is known to Cozy (for example Drive or Banks), and
the user has installed it on its device, the native app will
be opened.

The app's manifest can be set in the `app` prop. Then, in a
ReactNative environment, the AppLinker will be able to send
`openApp` message to the native environment with this `app`
data. Ideally the `mobile` member should be set with all data
needed to open the native app ([more info](https://github.com/cozy/cozy-stack/blob/master/docs/apps.md#mobile))

Handles several cases:

* On mobile app and other mobile app available
* On web (not mobile)
* On web mobile

As it uses the render props pattern, it is flexible and can be used to build components that are more complex than a simple
anchor.

```jsx
import AppLinker from 'cozy-ui/transpiled/react/AppLinker';

const app = {
  slug: 'banks'
};

<AppLinker app={app} href='http://dalailama-banks.mycozy.cloud'>{
  ({ onClick, href, name }) => (
    <a href={href} onClick={onClick}>
      Open { name }
    </a>
  )
}</AppLinker>
```

### Exemple with mobile data

```jsx
import AppLinker from 'cozy-ui/transpiled/react/AppLinker';

const app = {
  slug: 'passwords',
  mobile: {
    schema: 'cozypass://',
    id_playstore: 'io.cozy.pass',
    id_appstore: 'cozy-pass/id1502262449'
  }
};

<AppLinker app={app} href='http://dalailama-passwords.mycozy.cloud'>{
  ({ onClick, href, name }) => (
    <a href={href} onClick={onClick}>
      Open { name }
    </a>
  )
}</AppLinker>
```
