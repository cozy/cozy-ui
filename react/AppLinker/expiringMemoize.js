export default function(fn, duration, keyFn) {
  const memo = {}
  return arg => {
    const key = keyFn(arg)
    const memoInfo = memo[key]
    const uptodate =
      memoInfo && memoInfo.result && memoInfo.date - Date.now() < duration
    if (!uptodate) {
      memo[key] = { result: fn(arg), date: Date.now() }
    }
    return memo[key].result
  }
}
