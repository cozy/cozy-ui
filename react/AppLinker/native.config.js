export const UNIVERSAL_LINK_URL = 'https://links.mycozy.cloud'
