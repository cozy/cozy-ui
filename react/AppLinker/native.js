import { UNIVERSAL_LINK_URL } from 'cozy-ui/transpiled/react/AppLinker/native.config.js'
export const getUniversalLinkDomain = () => {
  return UNIVERSAL_LINK_URL
}
/*
  slug string drive 
  path : /path/to/view
  fallbackUrl: string : https://...mycozy.cloud
*/
export const generateUniversalLink = ({ slug, nativePath, fallbackUrl }) => {
  if (!nativePath.startsWith('/')) nativePath = '/' + nativePath
  let url = getUniversalLinkDomain() + '/' + slug + nativePath
  const urlObj = new URL(url)
  urlObj.searchParams.append('fallback', fallbackUrl)
  return urlObj.toString()
}
