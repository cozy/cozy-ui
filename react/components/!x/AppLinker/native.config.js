import { isAndroidApp } from 'cozy-device-helper'

export const NATIVE_APP_INFOS = {
  drive: {
    appId: 'io.cozy.drive.mobile',
    uri: 'cozydrive://',
    name: 'Cozy Drive'
  },
  banks: {
    appId: isAndroidApp() ? 'io.cozy.banks.mobile' : 'io.cozy.banks',
    uri: 'cozybanks://',
    name: 'Cozy Banks'
  }
}
export const UNIVERSAL_LINK_URL = 'https://links.mycozy.cloud'
