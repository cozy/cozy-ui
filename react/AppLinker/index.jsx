import React from 'react'
import PropTypes from 'prop-types'
import {
  checkApp,
  startApp,
  isMobileApp,
  isMobile,
  openDeeplinkOrRedirect,
  isAndroid
} from 'cozy-device-helper'
import {
  generateUniversalLink,
  generateWebLink,
  getUniversalLinkDomain
} from './native'

import { NATIVE_APP_INFOS } from 'cozy-ui/transpiled/react/AppLinker/native.config'
import expiringMemoize from './expiringMemoize'

const expirationDelay = 10 * 1000
const memoizedCheckApp = expiringMemoize(
  appInfo => checkApp(appInfo).catch(() => false),
  expirationDelay,
  appInfo => appInfo.appId
)

export class AppLinker extends React.Component {
  state = {
    nativeAppIsAvailable: null,
    isFetchingAppInfo: false
  }

  constructor(props) {
    super(props)
  }

  componentDidMount() {
    if (isMobileApp()) {
      this.checkAppAvailability()
    }
  }

  async checkAppAvailability() {
    const { slug } = this.props
    const appInfo = NATIVE_APP_INFOS[slug]
    if (appInfo) {
      const nativeAppIsAvailable = Boolean(await memoizedCheckApp(appInfo))
      this.setState({ nativeAppIsAvailable })
    }
  }

  static getOnClickHref(props, nativeAppIsAvailable) {
    const { slug, nativePath } = props
    let href = props.href
    let onClick = null
    const usingNativeApp = isMobileApp()
    const appInfo = NATIVE_APP_INFOS[slug]
    if (usingNativeApp) {
      if (nativeAppIsAvailable) {
        // If we are on the native app and the other native app is available,
        // we open the native app
        onClick = AppLinker.openNativeFromNative.bind(this, props)
        href = '#'
      } else {
        // If we are on a native app, but the other native app is not available
        // we open the web link, this is done by the href prop. We still
        // have to call the prop callback
        onClick = AppLinker.openWeb.bind(this, props)
      }
    } else if (isMobile() && appInfo) {
      // If we are on the "mobile web version", we try to open the native app
      // if it exists with an universal links. If it fails, we redirect to the web
      // version of the requested app
      // Only on iOS ATM
      if (isAndroid()) {
        onClick = AppLinker.openNativeFromWeb.bind(this, props)
      } else {
        //Since generateUniversalLink can rise an error, let's catch it to not crash
        //all the page.
        try {
          href = generateUniversalLink({ slug, nativePath, fallbackUrl: href })
        } catch (err) {
          console.error(err)
          href = '#'
        }
      }
    }

    return {
      href,
      onClick
    }
  }
  static openNativeFromWeb(props, ev) {
    const { href, slug, nativePath, onAppSwitch } = props
    const appInfo = NATIVE_APP_INFOS[slug]

    if (ev) {
      ev.preventDefault()
    }

    AppLinker.onAppSwitch(onAppSwitch)
    openDeeplinkOrRedirect(
      appInfo.uri + (nativePath === '/' ? '' : nativePath),
      function() {
        window.location.href = href
      }
    )
  }

  static onAppSwitch(onAppSwitchFn) {
    if (typeof onAppSwitchFn === 'function') {
      onAppSwitchFn()
    }
  }

  static openNativeFromNative(props, ev) {
    const { slug, onAppSwitch } = props
    if (ev) {
      ev.preventDefault()
    }
    const appInfo = NATIVE_APP_INFOS[slug]
    AppLinker.onAppSwitch(onAppSwitch)
    startApp(appInfo).catch(err => {
      console.error('AppLinker: Could not open native app', err)
    })
  }

  static openWeb(props) {
    AppLinker.onAppSwitch(props.onAppSwitch)
  }

  render() {
    const { children, slug } = this.props
    const { nativeAppIsAvailable } = this.state
    const appInfo = NATIVE_APP_INFOS[slug]
    const { href, onClick } = AppLinker.getOnClickHref(
      this.props,
      nativeAppIsAvailable
    )
    return children({ ...appInfo, onClick: onClick, href })
  }
}

AppLinker.defaultProps = {
  nativePath: '/'
}
AppLinker.propTypes = {
  //Slug of the app : drive / banks ...
  slug: PropTypes.string.isRequired,
  /*
  Full web url : Used by default on desktop browser
  Used as a fallback_uri on mobile web
  */
  href: PropTypes.string.isRequired,
  /*
    Path used for "native link"
  */
  nativePath: PropTypes.string,
  onAppSwitch: PropTypes.func
}

export default AppLinker
export {
  NATIVE_APP_INFOS,
  getUniversalLinkDomain,
  generateWebLink,
  generateUniversalLink
}
