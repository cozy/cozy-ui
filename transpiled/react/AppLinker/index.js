import _asyncToGenerator from "@babel/runtime/helpers/asyncToGenerator";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import _regeneratorRuntime from "@babel/runtime/regenerator";

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import PropTypes from 'prop-types';
import React from 'react';
import { withClient } from 'cozy-client';
import { checkApp, startApp, isMobileApp, isMobile, openDeeplinkOrRedirect, isAndroid, isFlagshipApp } from 'cozy-device-helper';
import { WebviewContext } from 'cozy-intent';
import logger from 'cozy-logger';
import expiringMemoize from "cozy-ui/transpiled/react/AppLinker/expiringMemoize";
import { generateUniversalLink, generateWebLink, getUniversalLinkDomain } from "cozy-ui/transpiled/react/AppLinker/native";
import { NATIVE_APP_INFOS } from "cozy-ui/transpiled/react/AppLinker/native.config";
var expirationDelay = 10 * 1000;
var memoizedCheckApp = expiringMemoize(function (appInfo) {
  return checkApp(appInfo).catch(function () {
    return false;
  });
}, expirationDelay, function (appInfo) {
  return appInfo.appId;
});
export var AppLinker = /*#__PURE__*/function (_React$Component) {
  _inherits(AppLinker, _React$Component);

  var _super = _createSuper(AppLinker);

  function AppLinker(props) {
    var _this;

    _classCallCheck(this, AppLinker);

    _this = _super.call(this, props);

    _defineProperty(_assertThisInitialized(_this), "state", {
      nativeAppIsAvailable: null,
      isFetchingAppInfo: false
    });

    _defineProperty(_assertThisInitialized(_this), "setImgRef", function (img) {
      _this.imgRef = img;

      _this.setState({
        imgRef: _this.imgRef
      });
    });

    _this.imgRef = null;
    return _this;
  }

  _createClass(AppLinker, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (isMobileApp()) {
        this.checkAppAvailability();
      }
    }
  }, {
    key: "checkAppAvailability",
    value: function () {
      var _checkAppAvailability = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var slug, appInfo, nativeAppIsAvailable;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                slug = AppLinker.getSlug(this.props);
                appInfo = NATIVE_APP_INFOS[slug];

                if (!appInfo) {
                  _context.next = 9;
                  break;
                }

                _context.t0 = Boolean;
                _context.next = 6;
                return memoizedCheckApp(appInfo);

              case 6:
                _context.t1 = _context.sent;
                nativeAppIsAvailable = (0, _context.t0)(_context.t1);
                this.setState({
                  nativeAppIsAvailable: nativeAppIsAvailable
                });

              case 9:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function checkAppAvailability() {
        return _checkAppAvailability.apply(this, arguments);
      }

      return checkAppAvailability;
    }()
  }, {
    key: "render",
    value: function render() {
      var children = this.props.children;
      AppLinker.deprecateSlug(this.props);
      var slug = AppLinker.getSlug(this.props);
      var nativeAppIsAvailable = this.state.nativeAppIsAvailable;
      var appInfo = NATIVE_APP_INFOS[slug];

      var _AppLinker$getOnClick = AppLinker.getOnClickHref(this.props, nativeAppIsAvailable, this.context, this.state.imgRef),
          href = _AppLinker$getOnClick.href,
          onClick = _AppLinker$getOnClick.onClick;

      return children(_objectSpread(_objectSpread({}, appInfo), {}, {
        iconRef: this.setImgRef,
        onClick: onClick,
        href: href
      }));
    }
  }], [{
    key: "getSlug",
    value: function getSlug(props) {
      if (props.app && props.app.slug) {
        return props.app.slug;
      }

      return props.slug;
    }
  }, {
    key: "deprecateSlug",
    value: function deprecateSlug(props) {
      if (props.slug) {
        console.warn("AppLinker's 'slug' prop is deprecated, please use 'app.slug' instead");
      }
    }
  }, {
    key: "getOnClickHref",
    value: function getOnClickHref(props, nativeAppIsAvailable, context, imgRef) {
      var app = props.app,
          client = props.client,
          nativePath = props.nativePath;
      var slug = AppLinker.getSlug(props);
      var href = props.href;
      var onClick = null;
      var usingNativeApp = isMobileApp();
      var appInfo = NATIVE_APP_INFOS[slug];

      if (isFlagshipApp()) {
        var _ref = client ? client.getInstanceOptions() : undefined,
            currentApp = _ref.app;

        if (currentApp === undefined || app.slug !== currentApp.slug) {
          var imgPayload = imgRef && JSON.stringify(_objectSpread({}, imgRef.getBoundingClientRect().toJSON()));
          return {
            onClick: function onClick(event) {
              event.preventDefault();
              context ? context.call('openApp', href, app, imgPayload) : logger.error("Failed to \"openApp(".concat(app, ")\". WebviewService has the following falsy value \"").concat(context, "\" in AppLinker's context."));
            },
            href: '#'
          };
        }
      }

      if (usingNativeApp) {
        if (nativeAppIsAvailable) {
          // If we are on the native app and the other native app is available,
          // we open the native app
          onClick = AppLinker.openNativeFromNative.bind(this, props);
          href = '#';
        } else {
          // If we are on a native app, but the other native app is not available
          // we open the web link, this is done by the href prop. We still
          // have to call the prop callback
          onClick = AppLinker.openWeb.bind(this, props);
        }
      } else if (isMobile() && appInfo) {
        // If we are on the "mobile web version", we try to open the native app
        // if it exists with an universal links. If it fails, we redirect to the web
        // version of the requested app
        // Only on iOS ATM
        if (isAndroid()) {
          onClick = AppLinker.openNativeFromWeb.bind(this, props);
        } else {
          // Since generateUniversalLink can rise an error, let's catch it to not crash
          // all the page.
          try {
            href = generateUniversalLink({
              slug: slug,
              nativePath: nativePath,
              fallbackUrl: href
            });
          } catch (err) {
            console.error(err);
            href = '#';
          }
        }
      }

      return {
        href: href,
        onClick: onClick
      };
    }
  }, {
    key: "openNativeFromWeb",
    value: function openNativeFromWeb(props, ev) {
      var href = props.href,
          nativePath = props.nativePath,
          onAppSwitch = props.onAppSwitch;
      var slug = AppLinker.getSlug(props);
      var appInfo = NATIVE_APP_INFOS[slug];

      if (ev) {
        ev.preventDefault();
      }

      AppLinker.onAppSwitch(onAppSwitch);
      openDeeplinkOrRedirect(appInfo.uri + (nativePath === '/' ? '' : nativePath), function () {
        window.location.href = href;
      });
    }
  }, {
    key: "onAppSwitch",
    value: function onAppSwitch(onAppSwitchFn) {
      if (typeof onAppSwitchFn === 'function') {
        onAppSwitchFn();
      }
    }
  }, {
    key: "openNativeFromNative",
    value: function openNativeFromNative(props, ev) {
      var onAppSwitch = props.onAppSwitch;
      var slug = AppLinker.getSlug(props);

      if (ev) {
        ev.preventDefault();
      }

      var appInfo = NATIVE_APP_INFOS[slug];
      AppLinker.onAppSwitch(onAppSwitch);
      startApp(appInfo).catch(function (err) {
        console.error('AppLinker: Could not open native app', err);
      });
    }
  }, {
    key: "openWeb",
    value: function openWeb(props) {
      AppLinker.onAppSwitch(props.onAppSwitch);
    }
  }]);

  return AppLinker;
}(React.Component);

_defineProperty(AppLinker, "contextType", WebviewContext);

AppLinker.defaultProps = {
  nativePath: '/'
};
AppLinker.propTypes = {
  /** DEPRECATED: please use app.slug prop */
  slug: PropTypes.string,

  /*
  Full web url : Used by default on desktop browser
  Used as a fallback_uri on mobile web
  */
  href: PropTypes.string,

  /*
    Path used for "native link"
  */
  nativePath: PropTypes.string,
  onAppSwitch: PropTypes.func,
  app: PropTypes.shape({
    // Slug of the app : drive / banks ...
    slug: PropTypes.string.isRequired,
    // Information about mobile native app
    mobile: PropTypes.shape({
      schema: PropTypes.string,
      id_playstore: PropTypes.string,
      id_appstore: PropTypes.string
    })
  }).isRequired
};
export default withClient(AppLinker);
export { NATIVE_APP_INFOS, getUniversalLinkDomain, generateWebLink, generateUniversalLink };