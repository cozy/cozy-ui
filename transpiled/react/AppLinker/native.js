import _defineProperty from "@babel/runtime/helpers/defineProperty";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
var _excluded = ["nativePath"];

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

import { generateWebLink as generateWebLinkClient, ensureFirstSlash } from 'cozy-client';
import { UNIVERSAL_LINK_URL } from "cozy-ui/transpiled/react/AppLinker/native.config";
export var getUniversalLinkDomain = function getUniversalLinkDomain() {
  return UNIVERSAL_LINK_URL;
};
export var generateWebLink = function generateWebLink(_ref) {
  var nativePath = _ref.nativePath,
      props = _objectWithoutProperties(_ref, _excluded);

  console.warn('Deprecated: you should use generateWebLink from cozy-client instead');
  return generateWebLinkClient(_objectSpread({
    pathname: '/',
    hash: nativePath
  }, props));
};
/**
 * Returns a universal link for an app + native path
 *
 * @param  {string} options.slug          - eg: drive
 * @param  {string} options.nativePath    - /path/to/view
 * @param  {string} options.fallbackUrl   - https://...mycozy.cloud, optional if cozyUrl is passed
 * @param  {string} options.cozyUrl       - https://name.mycozy.cloud, optional if fallbackUrl is passed
 * @param  {string} options.subDomainType - flat/nested, optionally the type of subdomains that is used.
 * @return {string}                       - https://links.cozy.cloud/drive/?fallback...
 */

export var generateUniversalLink = function generateUniversalLink(options) {
  var slug = options.slug,
      cozyUrl = options.cozyUrl,
      subDomainType = options.subDomainType;
  var fallbackUrl = options.fallbackUrl,
      nativePath = options.nativePath;
  nativePath = ensureFirstSlash(nativePath);

  if (!cozyUrl && !fallbackUrl) {
    throw new Error('Must have either cozyUrl or fallbackUrl to generate universal link.');
  }

  if (cozyUrl && !fallbackUrl) {
    fallbackUrl = generateWebLinkClient({
      cozyUrl: cozyUrl,
      pathname: '/',
      hash: nativePath,
      slug: slug,
      subDomainType: subDomainType
    });
  }

  var url = getUniversalLinkDomain(cozyUrl) + '/' + slug + nativePath;
  var urlObj = new URL(url);
  urlObj.searchParams.append('fallback', fallbackUrl);
  return urlObj.toString();
};