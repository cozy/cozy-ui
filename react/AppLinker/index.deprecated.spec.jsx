/**
 * Those tests are here to test previous AppLinker implementation base on
 * `slug` prop.
 * Now that AppLinkers implements `app` prop and uses `app.slug`, the old
 * `slug` prop is deprecated
 * Those tests should be kept until `slug` prop is completely removed
 */

import React from 'react'
import { shallow } from 'enzyme'
import {
  isMobileApp,
  isMobile,
  openDeeplinkOrRedirect,
  startApp,
  isAndroid
} from 'cozy-device-helper'

import AppLinker from './index'
import { generateUniversalLink } from './native'
jest.useFakeTimers()

const tMock = x => x

class DeprecatedAppItem extends React.Component {
  render() {
    const { app, onAppSwitch } = this.props
    return (
      <AppLinker
        onAppSwitch={onAppSwitch}
        slug={app.slug}
        href={'https://fake.link'}
      >
        {({ onClick, href, name }) => (
          <div>
            <a href={href} onClick={onClick}>
              Open {name}
            </a>
          </div>
        )}
      </AppLinker>
    )
  }
}
jest.mock('./native', () => ({
  ...jest.requireActual('./native'),
  generateUniversalLink: jest.fn()
}))

jest.mock('cozy-device-helper', () => ({
  ...jest.requireActual('cozy-device-helper'),
  isMobileApp: jest.fn(),
  isMobile: jest.fn(),
  openDeeplinkOrRedirect: jest.fn(),
  startApp: jest.fn().mockResolvedValue(),
  isAndroid: jest.fn()
}))

const app = {
  slug: 'drive',
  name: 'Drive'
}

describe('app icon', () => {
  let spyConsoleError, spyConsoleWarn, openNativeFromNativeSpy, appSwitchMock

  beforeEach(() => {
    isMobileApp.mockReturnValue(false)
    spyConsoleError = jest.spyOn(console, 'error')
    spyConsoleError.mockImplementation(() => {})
    spyConsoleWarn = jest.spyOn(console, 'warn')
    spyConsoleWarn.mockImplementation(() => {})
    openNativeFromNativeSpy = jest.spyOn(AppLinker, 'openNativeFromNative')
    isMobileApp.mockReturnValue(false)
    isMobile.mockReturnValue(false)
    isAndroid.mockReturnValue(false)
    appSwitchMock = jest.fn()
  })

  afterEach(() => {
    spyConsoleError.mockRestore()
    spyConsoleWarn.mockRestore()
    jest.restoreAllMocks()
  })

  it('should render correctly', () => {
    const root = shallow(<DeprecatedAppItem t={tMock} app={app} />).dive()
    expect(root.getElement()).toMatchSnapshot()
  })

  it('should work for native -> native', () => {
    const root = shallow(
      <DeprecatedAppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    ).dive()
    root.find('a').simulate('click')
    expect(appSwitchMock).not.toHaveBeenCalled()
    isMobileApp.mockReturnValue(true)
    root.setState({ nativeAppIsAvailable: true })
    root.find('a').simulate('click')
    expect(openNativeFromNativeSpy).toHaveBeenCalled()
    expect(startApp).toHaveBeenCalledWith({
      appId: 'io.cozy.drive.mobile',
      name: 'Cozy Drive',
      uri: 'cozydrive://'
    })
    expect(appSwitchMock).toHaveBeenCalled()
  })

  it('should work for web -> native for Android (custom schema)', () => {
    isMobile.mockReturnValue(true)
    isAndroid.mockResolvedValue(true)
    const root = shallow(
      <DeprecatedAppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    ).dive()
    root.find('a').simulate('click', { preventDefault: () => {} })
    expect(openDeeplinkOrRedirect).toHaveBeenCalledWith(
      'cozydrive://',
      expect.any(Function)
    )
    expect(appSwitchMock).toHaveBeenCalled()
  })

  it('should work for web -> native for iOS (universal link)', () => {
    isMobile.mockReturnValue(true)
    const root = shallow(
      <DeprecatedAppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    ).dive()
    root.find('a').simulate('click', { preventDefault: () => {} })

    expect(generateUniversalLink).toHaveBeenCalled()
  })

  it('should work for native -> web', () => {
    isMobileApp.mockReturnValue(true)
    const root = shallow(
      <DeprecatedAppItem t={tMock} app={app} onAppSwitch={appSwitchMock} />
    ).dive()
    root.find('a').simulate('click')
    expect(appSwitchMock).toHaveBeenCalled()
  })

  it('should not crash if no href', () => {
    isMobileApp.mockReturnValue(true)
    const root = shallow(
      <AppLinker onAppSwitch={appSwitchMock} slug={app.slug}>
        {({ onClick, href, name }) => (
          <div>
            <a href={href} onClick={onClick}>
              Open {name}
            </a>
          </div>
        )}
      </AppLinker>
    )
    expect(root.getElement()).toMatchSnapshot()
  })
})
