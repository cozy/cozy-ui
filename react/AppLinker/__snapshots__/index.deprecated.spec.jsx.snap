// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app icon should not crash if no href 1`] = `
<div>
  <a
    onClick={[Function]}
  >
    Open 
    Cozy Drive
  </a>
</div>
`;

exports[`app icon should render correctly 1`] = `
<div>
  <a
    href="https://fake.link"
    onClick={null}
  >
    Open 
    Cozy Drive
  </a>
</div>
`;
